File: backend/cache.py
import time
from typing import Any

_store: dict[str, tuple[Any, float]] = {}


def cache_get(key: str) -> Any | None:
    entry = _store.get(key)
    if entry and time.monotonic() < entry[1]:
        return entry[0]
    _store.pop(key, None)
    return None


def cache_set(key: str, value: Any, ttl: int = 30) -> None:
    _store[key] = (value, time.monotonic() + ttl)


def cache_invalidate(prefix: str) -> None:
    keys = [k for k in _store if k.startswith(prefix)]
    for k in keys:
        _store.pop(k, None)

File: backend-node/src/cache.ts
// In-memory TTL cache — mirrors backend/cache.py (per-process, monotonic TTL).
interface Entry {
  value: unknown;
  expiresAt: number;
}

const store = new Map<string, Entry>();

export function cacheGet<T = unknown>(key: string): T | null {
  const entry = store.get(key);
  if (entry && Date.now() < entry.expiresAt) return entry.value as T;
  store.delete(key);
  return null;
}

export function cacheSet(key: string, value: unknown, ttlSeconds = 30): void {
  store.set(key, { value, expiresAt: Date.now() + ttlSeconds * 1000 });
}

export function cacheInvalidate(prefix: string): void {
  for (const key of store.keys()) {
    if (key.startsWith(prefix)) store.delete(key);
  }
}
